Makefile

```diff
@@ -37,28 +49,60 @@ cluster-destroy:
 
 .PHONY: spark-build
 spark-build:
-	if [[ -z  "$(shell docker images -q $(SPARK_IMAGE_NAME_WITH_TAG))" ]]; then
-		docker build -t ${SPARK_IMAGE_NAME_WITH_TAG} ${SPARK_IMAGE_DOCKERFILE_PATH}
+	if [[ -z  "$(shell docker images -q $(SPARK_IMAGE_FULL_NAME))" ]]; then
+		docker build \
+			-t ${SPARK_IMAGE_FULL_NAME} \
+			-f ${SPARK_IMAGE_DOCKERFILE_PATH}/Dockerfile \
+			${SPARK_IMAGE_DOCKERFILE_PATH}
 	fi
-	echo "${SPARK_IMAGE_NAME_WITH_TAG}" > $@
+	echo "${SPARK_IMAGE_FULL_NAME}" > $@
 
 .PHONY: docker-build
 docker-build: spark-build
 docker-build:
-	docker build \
-		--build-arg SPARK_IMAGE=$(shell cat spark-build) \
-		-t ${SPARK_ON_K8S_OPERATOR_IMAGE_NAME} \
-		-f ${SPARK_ON_K8S_OPERATOR_DOCKERFILE_PATH} ${SPARK_OPERATOR_DIR}
+	if [[ -z "$(shell docker images -q $(SPARK_ON_K8S_OPERATOR_IMAGE_FULL_NAME))" ]]; then
+		docker build \
+			--build-arg SPARK_IMAGE=$(shell cat spark-build) \
+			-t ${SPARK_ON_K8S_OPERATOR_IMAGE_FULL_NAME} \
+			-f ${SPARK_ON_K8S_OPERATOR_DOCKERFILE_PATH}/Dockerfile \
+			${SPARK_OPERATOR_DIR}
+		# clean-up builder stage
+		docker image prune -f --filter label=stage=spark-operator-builder
+	fi
 
 .PHONY: docker-push
 docker-push:
-	docker push ${SPARK_ON_K8S_OPERATOR_IMAGE_NAME}
+	docker push ${SPARK_ON_K8S_OPERATOR_IMAGE_FULL_NAME}
+
+.PHONY: docker-builder
+docker-builder:
+	if [[ -z "$(shell docker images -q $(DOCKER_BUILDER_IMAGE_FULL_NAME))" ]]; then
+		docker build \
+			-t $(DOCKER_BUILDER_IMAGE_FULL_NAME) \
+			-f ${DOCKER_BUILDER_DOCKERFILE_PATH}/Dockerfile \
+			${DOCKER_BUILDER_DOCKERFILE_PATH}
+	fi
+	echo $(DOCKER_BUILDER_IMAGE_FULL_NAME) > $@
 
+.PHONY: test
+test: docker-builder
 test:
-	$(ROOT_DIR)/run-tests.sh
+	docker run -i --rm \
+		-v $(ROOT_DIR)/tests:/tests \
+		-v $(ROOT_DIR)/run-tests.sh:/run-tests.sh \
+		-v $(ROOT_DIR)/admin.conf:/root/.kube/config \
+		-e SPARK_IMAGE=$(SPARK_IMAGE_FULL_NAME) \
+		-e SPARK_OPERATOR_IMAGE=$(SPARK_ON_K8S_OPERATOR_IMAGE_FULL_NAME) \
+		$(shell cat docker-builder) \
+		/bin/bash -c \
+		"kubectl config use-context kubernetes-admin@kudo-spark-operator && \
+		kubectl cluster-info && \
+		echo \$$SPARK_IMAGE && \
+		echo \$$SPARK_OPERATOR_IMAGE && \
+		/run-tests.sh"
```

Comment:
will that work well with CI in terms of all the environment variables passed to  `/run-tests.sh`?  We should either invoke `make test` from `/run-tests.sh` and do all the configuration here, or make sure CI will set all the same parameters needed for the script. The contract is that CI triggers only `/run-tests.sh`